File: src/main/java/dominio/empresas/LectorArchivos.java
package dominio.empresas;

import excepciones.NoSePudoLeerEseTipoDeArchivoError;

import java.util.ArrayList;
import java.util.List;

public class LectorArchivos {
	private List<ArchivoEmpresas> lectores = new ArrayList<>();//Se inicializa en el constructor
	
	public LectorArchivos(String ruta){
		lectores.add(new ArchivoXLS(ruta));
		lectores.add(new ArchivoCSV(ruta));
	}
	
	public List<Empresa> getEmpresas(){
		ArchivoEmpresas archivo = obtenerLectorApropiado();
		archivo.leerEmpresas();
		return archivo.getEmpresas();
	}
	
	private ArchivoEmpresas obtenerLectorApropiado(){
		return lectores.stream().filter(lector -> lector.puedeLeerArchivo()).findFirst().orElseThrow(() -> new NoSePudoLeerEseTipoDeArchivoError("No se pudo leer el archivo."));
	}

}

File: src/main/java/dondeInvierto/dominio/empresas/LectorArchivos.kt
package dondeInvierto.dominio.empresas

import excepciones.NoSePudoLeerEseTipoDeArchivoError

class LectorArchivos(ruta: String) {
    private val lectores = listOf(ArchivoXLS(ruta), ArchivoCSV(ruta))

    fun getEmpresas(): List<Empresa> {
        val archivo = obtenerLectorApropiado()
        archivo.leerEmpresas()
        return archivo.getEmpresas()
    }

    private fun obtenerLectorApropiado(): ArchivoEmpresas {
        return lectores.firstOrNull { it.puedeLeerArchivo() }
                ?: throw NoSePudoLeerEseTipoDeArchivoError("No se pudo leer el archivo.")
    }

}